keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,897 @@
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+- ((TODO))
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### ALTERNATIVE 1: Pod-Deletion-Cost annotation
+
+Same as the current [`controller.kubernetes.io/pod-deletion-cost`](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) annotation.
+
+__USAGE PATTERN 1 - the annotation kept up-to-date with a controller:__
+
+- PROBLEM 1: every update requires a PATCH call to the kube-apiserver (probably overloading it)
+- PROBLEM 2: this is wasteful for deployments that rarely scale down
+   - calculating and updating the cost could be expensive, and that work is wasted if not actually used to downscale
+
+__USAGE PATTERN 2 - the annotation is updated only when downscaling:__
+
+- PROBLEM 1: existing scaling tools like HorozontalPodAutoscaler can't be used
+   - the annotation/status would need to be updated BEFORE downscaling, and we can't predict when HorozontalPodAutoscaler will downscale
+   - therefore, users must write their own scalers that update the annotation/status before downscaling (making it inaccessible for most users)
+- PROBLEM 2: after each downscale, any annotations added by the scaler will need to be cleared (they will become out of date)
+   - related to this, we must wait to clear them before we can start the next downscale (or the old annotations will impact the new scale)
+- PROBLEM 3: costs may be out-of-date by the time the controller picks which pod to downscale
+   - some apps like airflow/spark may have quite rapidly changing "best" pods to kill
+
+### ALTERNATIVE 2: Pod-Deletion-Cost http/exec probe
+
+A new http/exec probe is created for Pods which returns their current pod-deletion-cost.
+
+__USAGE PATTERN 1 - probes are used every X seconds to update a Pod status field:__
+
+- (Suffers from the same problems as "USAGE PATTERN 1" of "ALTERNATIVE 1")
+
+__USAGE PATTERN 2 - probes are ONLY used when downscaling:__
+
+- PROBLEM 1: the controller cannot make a probe request to Pods (this must be done by the Node's kubelet)
+  - to solve this you would need a complex system that has the controller "mark" the pods for probing by the kubelet
+- PROBLEM 2: to find the "best" Pod to kill, we would need to probe every single Pod, this is not scalable
+  - to solve this, you would need to use a heuristic approach, e.g. only checking a sample of Pods and returning the lowest cost from the sample
+
+### ALTERNATIVE 3: Pod-Deletion-Cost API
+
+A central user-managed API is queried by the controller and returns the pod-deletion-cost of each Pod.
+(Rather than returning a list of `chosen_pods` and `tied_pods` like in this KEP)
```

Comment:
It could have this instead.

Reply:
Can you clarify what you mean "this", how would you address the problems listed?